docs/picchick.1
'\" t
.\"     Title: picchick
.\"    Author: [see the "AUTHOR(S)" section]
.\" Generator: Asciidoctor 2.0.23
.\"      Date: 2025-09-09
.\"    Manual: Picchick Manual
.\"    Source: picchick v0.4.99-dev
.\"  Language: English
.\"
.TH "PICCHICK" "1" "2025-09-09" "picchick v0.4.99\-dev" "Picchick Manual"
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.ss \n[.ss] 0
.nh
.ad l
.de URL
\fI\\$2\fP <\\$1>\\$3
..
.als MTO URL
.if \n[.g] \{\
.  mso www.tmac
.  am URL
.    ad l
.  .
.  am MTO
.    ad l
.  .
.  LINKSTYLE blue R < >
.\}
.SH "NAME"
picchick \- A utility for programming and debugging microcontrollers.
.SH "SYNOPSIS"
.sp
\fBpicchick [\-d <mcu>] [\-c <programmer>] [\-r <addr> [len] | \-w <addr> <word> | \-f] [\-e [addr]] [\-v] [hexfile]\fP
.br
\fBpicchick [\-d <mcu>] \-\-map <hexfile>\fP
.br
\fBpicchick \-\-list\-devices\fP
.sp
\fBpicchick [\-c <programmer>] \-h/\-\-help\fP
.SH "OPTIONS"
.sp
picchick accepts some common configuration options that dictate how to go about
performing the action. Other options provide information about picchick itself.
The programmer flag is required for all actions that involve a programmer.
The device flag is required for all actions that involve a device.
.sp
\fB\-c <programmer>\fP
.RS 4
The name of the programmer to use. Different programmers
require different options, see the programmer\(cqs options with
\f(CRpicchick \-c <programmer> \-h\fP.
.RE
.sp
\fB\-d, \-\-device <mcu>\fP
.RS 4
The device being programmed. The \fImcu\fP argument is either
a chip part number, or a path to a custom device file.
.RE
.sp
\fB\-\-version\fP
.RS 4
Print the version number and exit.
.RE
.sp
\fB\-h, \-\-help\fP
.RS 4
Print usage information and a short help message. Can be
combined with \f(CR\-c\fP to get help about a specific programmer.
.RE
.SH "ACTIONS"
.sp
At least one action is required, else picchick would have nothing to do! Some
actions cannot be used together, i.e. you cannot flash and write in the same
command. Other actions can be chained, for instance you can erase a device,
flash a hexfile, then verify the memory in one command.
.sp
\fB\-r, \-\-read <addr> [len]\fP
.RS 4
Read \fIlength\fP words starting from \fIaddr\fP. Length
defaults to 2 if omitted.
.RE
.sp
\fB\-w, \-\-write <addr> <word>\fP
.RS 4
Write \fIword\fP to \fIaddress\fP.
.RE
.sp
\fB\-f, \-\-flash\fP
.RS 4
Flash \fIhexfile\fP to device.
.RE
.sp
\fB\-e, \-\-erase [addr]\fP
.RS 4
Erase \fIaddr\fP. Address can be a hex number, \*(Aqall\*(Aq or \*(Aqflash\*(Aq.
.RE
.sp
\fB\-v, \-\-verify\fP
.RS 4
Verify written data. This can be used with either the write or
flash actions, or by itself to verify against a hexfile.
.RE
.sp
\fB\-\-map\fP
.RS 4
Pretty print \fIhexfile\fP, mapped to \fIdevice\fP if the \f(CR\-d\fP flag is given.
.RE
.sp
\fB\-\-list\-ports\fP
.RS 4
List available serial ports.
.RE
.SH "PROGRAMMERS"
.SS "picstick"
.sp
picstick is used for PIC16F1 devices that implement the low\-voltage ICSP
interface.
.SS "flipflop"
.sp
flipflop is a bootloader that runs on PIC16F1 devices with an EUSART.
.SS "arduino\-icsp"
.sp
Arduino powered low\-voltage ICSP programmer for PICs.
.SH "RESOURCES"
.sp
\fBGithub:\fP \c
.URL "https://github.com/Rex\-\-/picchick" "" ""
.sp
\fBDocumentation:\fP \c
.URL "https://rex.mckinnon.ninja/picchick" "" ""
.SH "COPYING"
.sp
Copyright \(co 2022\-2025 Rex McKinnon
.br
This software is released under the University of Illinois/NCSA
Open Source License. Check the LICENSE file for more details.

sw/picchick.py
#! /bin/python

import argparse
import os.path
import sys

import hexfile
import programmer


DESCRIPTION = '''\
A utility for programming PIC19196 microcontrollers\
'''

USAGE = '''\
%(prog)s [options] [hexfile]\
'''

EPILOG = '''\
flag arguments:
  [addr]:\t\tdevice memory address in hexadecimal
\tall\t\tall device memory areas
\tflash\t\tuser flash area
\tconfig\t\tcofiguration words area
\tuserid\t\tuser identification area
\tdevid\t\tdevice identification area (read-only)
'''

parser = argparse.ArgumentParser(
    formatter_class=argparse.RawDescriptionHelpFormatter,
    description=DESCRIPTION,
    usage=USAGE,)

parser.add_argument('hexfile',
    nargs='?',
    default=None,
    help='path to the hexfile')

parser.add_argument('-f', '--flash',
    action='store_true',
    help='flash hexfile onto the device')
parser.add_argument('--read',
    metavar='addr',
    help='read specified address or chunk  of memory')
parser.add_argument('--write',
    nargs=2,
    metavar=('addr', 'word'),
    help='write word to specified address or chunk  of memory')
parser.add_argument('--erase',
    nargs='?',
    const='all',
    metavar='addr',
    help='erase device or specified address')

# parser.add_argument('-d', '--device',
#     metavar='dev',
#     help='device to be programmed')
parser.add_argument('-p', '--port',
    metavar='port',
    help='programmer serial port')
parser.add_argument('--baud',
    type=int,
    default=9600,
    metavar='baud',
    help='serial connection baudrate',)

parser.add_argument('--map',
    action='store_true',
    help='display the hexfile')
parser.add_argument('--list-ports',
    action='store_true',
    help='list available serial ports')
parser.add_argument('--list-devices',
    action='store_true',
    help='list available device configurations')


args = parser.parse_args()

# Requirements tree

# Flash flag requires both the hexfile and the programmer (port)
both_reqd = (args.flash)
# The read and erase flags only require the programmer connection (and port flag)
programmer_reqd = both_reqd or (args.read or args.erase or args.write)
# The map flag only requires the hexfile to be present
hexfile_reqd = both_reqd or (args.map)


# Firstly, if we need the hexfile, check if it exists
# If not, immediatly exit with a helpful message
if hexfile_reqd:
    if args.hexfile is None:
        print(f"Missing argument: hexfile")
        sys.exit(1)
    elif not os.path.isfile(args.hexfile):
        print(f"Could not find hexfile: { args.hexfile}")
        sys.exit(1)
    else:
        print(f"Using hexfile: { args.hexfile }")
    hex_decoder = hexfile.HexfileDecoder(args.hexfile)

# We now have all the hexfile reqs, so take care of the actions
# that only require the hexfile
if args.map:
    hex_decoder.printMemory()


# Second if we need the programmer, we check:
# - If the -p argument is specified
# - If the path exists (This may not work on windows)
if programmer_reqd:
    if args.port is None:
        print("Missing argument: -p port")
        sys.exit(1)
    elif not os.path.exists(args.port):
        print(f"Could not find port: { args.port }")
        sys.exit(1)
    else:
        dev = programmer.Programmer(args.port, baud=args.baud)
        if not dev.connect():
            print(f"ERROR: Failed to connect to device: { args.port } Exiting...")
            sys.exit(1)


# We now have all the programmer reqs, so do the actions that only
# need the programmer:
# Display information about ports if flag was included
if args.list_ports:
    # if args.port is None:
    programmer.listPorts()
    # else:
        # print("Not including port-list due to valid port flag")


if args.erase or args.flash or args.read or args.write:
    dev.start()

    if args.erase:
        if args.erase == 'all':
            dev.erase(0xFFFF)
        elif args.erase == 'flash':
            dev.erase(0xEFFF)
        else:
            dev.erase(int(args.erase, base=16))
    
    if args.flash:
        

        for start_address, row in hex_decoder.memory.items():
            if start_address < hexfile.USER_ID_START:
                dev.row(start_address, row)
        for start_address, word in hex_decoder.memory.items():
            if hexfile.USER_ID_START <= start_address < hexfile.CONFIG_WORD_START:
                dev.word(start_address, word[0])
        for start_address, word in hex_decoder.memory.items():
            if hexfile.CONFIG_WORD_START <= start_address:
                dev.word(start_address, word[0])
    elif args.write:
        dev.word(int(args.write[0], base=16), int(args.write[1], base=16))
    
    if args.read:
        dev.read(int(args.read, base=16))
    dev.stop() 
    dev.disconnect()
